File: pygennet/evolution.py
import numpy as np
from .models import NeuralNetwork

class Evolution:
    """
    Provides methods to run an evolutionary algorithm on NeuralNetworks.
    """

    @staticmethod
    def create_initial_population(population_size: int, network_architecture: list[int]) -> list[NeuralNetwork]:
        """
        Creates an initial population of random neural networks.
        """
        population = []
        for _ in range(population_size):
            network = NeuralNetwork(architecture=network_architecture)
            population.append(network)
        return population

    @staticmethod
    def select_fittest(population_with_fitness: list[tuple[NeuralNetwork, float]]) -> list[NeuralNetwork]:
        """
        Selects the top-performing half of the population based on fitness scores.
        'population_with_fitness' should be a list of (network, fitness_score) tuples.
        """
        # Sort networks by fitness in descending order (higher is better)
        sorted_population = sorted(population_with_fitness, key=lambda x: x[1], reverse=True)

        # Select the top half
        num_to_select = len(sorted_population) // 2
        fittest_networks = [item[0] for item in sorted_population[:num_to_select]]

        return fittest_networks

    @staticmethod
    def reproduce(fittest_networks: list[NeuralNetwork], new_population_size: int, mutation_rate: float, mutation_chance: float) -> list[NeuralNetwork]:
        """
        Creates a new generation by cloning and mutating the fittest networks.
        """
        new_population = []

        # Ensure we have some networks to reproduce from
        if not fittest_networks:
            # If the fittest list is empty, we can't reproduce.
            # This might happen if selection returns nothing.
            # We could either raise an error or return an empty list.
            return []

        while len(new_population) < new_population_size:
            # Choose a parent from the fittest networks at random
            parent = np.random.choice(fittest_networks)

            # Create a child by cloning the parent's parameters
            child = NeuralNetwork(architecture=parent.architecture)
            child.set_parameters(parent.get_parameters())

            # Mutate the child
            child.mutate(mutation_rate=mutation_rate, mutation_chance=mutation_chance)

            new_population.append(child)

        return new_population

File: src/evolution.c
#include "evolution.h"
#include <stdlib.h>
#include <stdio.h>

// --- Evolution Functions Implementation ---

// Creates an initial population of neural networks
NeuralNetwork** create_initial_population(int population_size, int num_layers, const int* architecture) {
    NeuralNetwork** population = (NeuralNetwork**)malloc(population_size * sizeof(NeuralNetwork*));
    if (!population) return NULL;

    for (int i = 0; i < population_size; i++) {
        population[i] = create_neural_network(num_layers, architecture);
    }
    return population;
}

// Comparison function for qsort to sort networks by fitness in descending order
int compare_fitness(const void* a, const void* b) {
    const NetworkFitness* nf_a = (const NetworkFitness*)a;
    const NetworkFitness* nf_b = (const NetworkFitness*)b;
    if (nf_a->fitness < nf_b->fitness) return 1;
    if (nf_a->fitness > nf_b->fitness) return -1;
    return 0;
}

// Selects the fittest networks from a population
NetworkFitness* select_fittest(NetworkFitness* population_with_fitness, int population_size, int* num_fittest) {
    // Sort the population by fitness
    qsort(population_with_fitness, population_size, sizeof(NetworkFitness), compare_fitness);

    // Select the top half
    *num_fittest = population_size / 2;
    NetworkFitness* fittest = (NetworkFitness*)malloc(*num_fittest * sizeof(NetworkFitness));
    if (!fittest) {
        *num_fittest = 0;
        return NULL;
    }

    for (int i = 0; i < *num_fittest; i++) {
        fittest[i] = population_with_fitness[i];
    }

    return fittest;
}

// Creates a new generation by cloning and mutating the fittest networks
NeuralNetwork** reproduce(const NetworkFitness* fittest_networks, int num_fittest, int new_population_size, float mutation_rate, float mutation_chance) {
    if (num_fittest == 0) return NULL;

    NeuralNetwork** new_population = (NeuralNetwork**)malloc(new_population_size * sizeof(NeuralNetwork*));
    if (!new_population) return NULL;

    for (int i = 0; i < new_population_size; i++) {
        // Choose a random parent from the fittest networks
        int parent_index = rand() % num_fittest;
        const NeuralNetwork* parent = fittest_networks[parent_index].network;

        // Clone the parent to create a child
        NeuralNetwork* child = clone_network(parent);

        // Mutate the child
        mutate_network(child, mutation_rate, mutation_chance);

        new_population[i] = child;
    }

    return new_population;
}
